File: go/5.go
// https://leetcode.com/problems/longest-palindromic-substring/

func longestPalindrome(s string) string {
    n := len(s)
    if n < 2 {
        return s
    }

    start, maxLength := 0, 1
    dp := make([][]bool, n)
    for i := range dp {
        dp[i] = make([]bool, n)
        dp[i][i] = true
    }

    for begin := n - 1; begin >= 0; begin-- {
        for end := begin + 1; end < n; end++ {
            if s[begin] == s[end] {
                if end-begin == 1 || dp[begin+1][end-1] {
                    dp[begin][end] = true
                    if end-begin+1 > maxLength {
                        start = begin
                        maxLength = end - begin + 1
                    }
                }
            }
        }
    }

    return s[start : start+maxLength]
}

File: rust/5.r
# https://leetcode.com/problems/longest-palindromic-substring/

impl Solution {
    pub fn longest_palindrome(s: String) -> String {
        let chars: Vec<char> = s.chars().collect();
        let len = chars.len();
        let mut dp = vec![vec![false; len]; len];
        let mut left = 0;
        let mut right = 0;

        for j in 0..len {
            for i in 0..=j {
                dp[i][j] = chars[i] == chars[j] && (j - i < 2 || dp[i + 1][j - 1]);
                if dp[i][j] && j + 1 - i > right - left {
                    left = i;
                    right = j + 1;
                }
            }
        }

        chars[left..right].iter().collect()
    }
}